src/ray/util/event.cc

```diff
@@ -16,11 +16,213 @@
+      stop_writing_ = true;
+      return;
+    } else {
+      file_length_ += message_len;
+      bytes_since_flush_ += message_len;
+    }
+  }
+
+  if (force_flush || bytes_since_flush_ > max_flush_size_) {
+    FlushUnlocked();
+  }
+}
+
+///
+/// LogEventReporter
+///
+LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type, std::string& log_dir)
+    : log_dir_(log_dir) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+
+  log_sink_ = std::unique_ptr<LogFileObject>(new LogFileObject(log_dir_, file_name_));
+}
+
+std::string LogEventReporter::EventToString(const rpc::Event &event) {
+  std::stringstream result;
+
+  boost::property_tree::ptree pt;
+  pt.put("time_stamp", event.timestamp());
+  pt.put("severity", Event_Severity_Name(event.severity()));
+  pt.put("label", event.label());
+  pt.put("event_id", event.event_id());
+  pt.put("source_type", Event_SourceType_Name(event.source_type()));
+  pt.put("host_name", event.source_hostname());
+  pt.put("pid", std::to_string(event.source_pid()));
+
+  std::stringstream message_buffer;
+  const std::string &message = event.message();
+  for (int i = 0, len = message.size(); i < len; ++i) {
+    if (message[i] == '\n' || message[i] == '\r') {
+      message_buffer << "\\n";
+    } else if (message[i] == '\"') {
+      message_buffer << "\'";
+    } else {
+      message_buffer << message[i];
+    }
+  }
+
+  pt.put("message", message_buffer.str());
+  for (auto &ele : event.custom_fields()) {
+    pt.put("custom_fields." + ele.first, ele.second);
+  }
+
+  std::stringstream ss;
+  boost::property_tree::json_parser::write_json(ss, pt, false);
+  return ss.str();
+}
+
+void LogEventReporter::Report(const rpc::Event &event) {
+  RAY_CHECK(Event_SourceType_IsValid(event.source_type()));
+  RAY_CHECK(Event_Severity_IsValid(event.severity()));
+  std::string result = EventToString(event);
+  log_sink_->Write(
+      result.c_str(), result.size(),
+      event.severity() == rpc::Event_Severity::Event_Severity_FATAL ? true : false);
+}
+///
 /// EventManager
 ///
+
+std::unique_ptr<EventManager> EventManager::instance_ = nullptr;
+absl::Mutex EventManager::mutex_;
+
 EventManager &EventManager::Instance() {
-  static EventManager instance_;
-  return instance_;
+  if (instance_ == nullptr) {
+    absl::MutexLock lock(&mutex_);
+    if (instance_ == nullptr) {
+      instance_ = std::unique_ptr<EventManager>(new EventManager());
+    }
+  }
+  return *instance_;
+}
+
+EventManager::~EventManager() {
+  for (const auto &element : reporter_map_) {
+    (element.second)->Close();
+  }
+  reporter_map_.clear();
+  io_work_.~work();
```

Comment:
Why call deconstructor here? It will be executed before end of EventManager deconstructor.